File: native/ios/VueNativeCore/Sources/VueNativeCore/Modules/SocialAuthModule.swift
#if canImport(UIKit)
import UIKit
import AuthenticationServices

/// Native module for social authentication (Apple Sign In, Google Sign In).
///
/// Methods:
///   - signInWithApple() -- present Apple Sign In sheet
///   - signInWithGoogle(clientId: String) -- present Google OAuth web flow
///   - signOut(provider: String) -- clear cached credentials
///   - getCurrentUser(provider: String) -- check for existing session
///
/// Events:
///   - auth:appleCredentialRevoked -- fired when Apple credential is revoked
final class SocialAuthModule: NSObject, NativeModule, ASAuthorizationControllerDelegate, ASAuthorizationControllerPresentationContextProviding {
    let moduleName = "SocialAuth"
    private weak var bridge: NativeBridge?
    private var appleSignInCallback: ((Any?, String?) -> Void)?
    private var credentialObserver: NSObjectProtocol?

    init(bridge: NativeBridge? = nil) {
        self.bridge = bridge
        super.init()
        observeAppleCredentialRevocation()
    }

    func invoke(method: String, args: [Any], callback: @escaping (Any?, String?) -> Void) {
        switch method {
        case "signInWithApple":
            handleAppleSignIn(callback: callback)

        case "signInWithGoogle":
            guard let clientId = args.first as? String else {
                callback(nil, "signInWithGoogle: expected clientId string")
                return
            }
            handleGoogleSignIn(clientId: clientId, callback: callback)

        case "signOut":
            let provider = args.first as? String ?? "apple"
            handleSignOut(provider: provider, callback: callback)

        case "getCurrentUser":
            let provider = args.first as? String ?? "apple"
            handleGetCurrentUser(provider: provider, callback: callback)

        default:
            callback(nil, "SocialAuthModule: unknown method '\(method)'")
        }
    }

    // MARK: - Apple Sign In

    private func handleAppleSignIn(callback: @escaping (Any?, String?) -> Void) {
        appleSignInCallback = callback

        let request = ASAuthorizationAppleIDProvider().createRequest()
        request.requestedScopes = [.fullName, .email]

        let controller = ASAuthorizationController(authorizationRequests: [request])
        controller.delegate = self
        controller.presentationContextProvider = self
        controller.performRequests()
    }

    // ASAuthorizationControllerDelegate
    func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
        guard let credential = authorization.credential as? ASAuthorizationAppleIDCredential else {
            appleSignInCallback?(nil, "signInWithApple: unexpected credential type")
            appleSignInCallback = nil
            return
        }

        var result: [String: Any] = [
            "userId": credential.user,
        ]

        if let email = credential.email {
            result["email"] = email
        }

        if let fullName = credential.fullName {
            let name = [fullName.givenName, fullName.familyName]
                .compactMap { $0 }
                .joined(separator: " ")
            if !name.isEmpty {
                result["fullName"] = name
            }
        }

        if let tokenData = credential.identityToken, let token = String(data: tokenData, encoding: .utf8) {
            result["identityToken"] = token
        }

        if let codeData = credential.authorizationCode, let code = String(data: codeData, encoding: .utf8) {
            result["authorizationCode"] = code
        }

        // Persist user ID for session check
        UserDefaults.standard.set(credential.user, forKey: "vn_apple_userId")

        appleSignInCallback?(result, nil)
        appleSignInCallback = nil
    }

    func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Error) {
        let nsError = error as NSError
        if nsError.domain == ASAuthorizationError.errorDomain && nsError.code == ASAuthorizationError.canceled.rawValue {
            appleSignInCallback?(nil, "signInWithApple: user cancelled")
        } else {
            appleSignInCallback?(nil, "signInWithApple: \(error.localizedDescription)")
        }
        appleSignInCallback = nil
    }

    func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        return UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first?.windows.first { $0.isKeyWindow } ?? UIWindow()
    }

    // MARK: - Google Sign In (URL-based OAuth)

    private func handleGoogleSignIn(clientId: String, callback: @escaping (Any?, String?) -> Void) {
        // URL-based OAuth flow using ASWebAuthenticationSession for zero dependencies
        let redirectURI = "com.googleusercontent.apps.\(clientId.components(separatedBy: ".").first ?? clientId):/oauthredirect"
        let scope = "openid email profile"
        let urlString = "https://accounts.google.com/o/oauth2/v2/auth?client_id=\(clientId)&redirect_uri=\(redirectURI)&response_type=code&scope=\(scope.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? scope)"

        guard let url = URL(string: urlString) else {
            callback(nil, "signInWithGoogle: invalid OAuth URL")
            return
        }

        DispatchQueue.main.async {
            let scheme = redirectURI.components(separatedBy: ":").first
            let session = ASWebAuthenticationSession(url: url, callbackURLScheme: scheme) { callbackURL, error in
                if let error = error {
                    let nsError = error as NSError
                    if nsError.domain == ASWebAuthenticationSessionError.errorDomain &&
                       nsError.code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
                        callback(nil, "signInWithGoogle: user cancelled")
                    } else {
                        callback(nil, "signInWithGoogle: \(error.localizedDescription)")
                    }
                    return
                }

                guard let callbackURL = callbackURL,
                      let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false),
                      let code = components.queryItems?.first(where: { $0.name == "code" })?.value else {
                    callback(nil, "signInWithGoogle: no authorization code received")
                    return
                }

                let result: [String: Any] = [
                    "userId": code, // The auth code — exchange server-side for tokens
                    "authorizationCode": code,
                ]

                UserDefaults.standard.set(code, forKey: "vn_google_authCode")
                callback(result, nil)
            }

            session.prefersEphemeralWebBrowserSession = false

            // Find the presentation anchor
            if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
               let window = windowScene.windows.first {
                session.presentationContextProvider = window.rootViewController as? ASWebAuthenticationPresentationContextProviding
            }

            session.start()
        }
    }

    // MARK: - Sign Out

    private func handleSignOut(provider: String, callback: @escaping (Any?, String?) -> Void) {
        switch provider {
        case "apple":
            UserDefaults.standard.removeObject(forKey: "vn_apple_userId")
        case "google":
            UserDefaults.standard.removeObject(forKey: "vn_google_authCode")
        default:
            break
        }
        callback(nil, nil)
    }

    // MARK: - Get Current User

    private func handleGetCurrentUser(provider: String, callback: @escaping (Any?, String?) -> Void) {
        switch provider {
        case "apple":
            guard let userId = UserDefaults.standard.string(forKey: "vn_apple_userId") else {
                callback(nil, nil)
                return
            }
            // Verify the credential is still valid
            ASAuthorizationAppleIDProvider().getCredentialState(forUserID: userId) { state, _ in
                switch state {
                case .authorized:
                    callback(["userId": userId], nil)
                default:
                    UserDefaults.standard.removeObject(forKey: "vn_apple_userId")
                    callback(nil, nil)
                }
            }
        case "google":
            if let code = UserDefaults.standard.string(forKey: "vn_google_authCode") {
                callback(["userId": code], nil)
            } else {
                callback(nil, nil)
            }
        default:
            callback(nil, nil)
        }
    }

    // MARK: - Credential Revocation Observer

    private func observeAppleCredentialRevocation() {
        credentialObserver = NotificationCenter.default.addObserver(
            forName: ASAuthorizationAppleIDProvider.credentialRevokedNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            UserDefaults.standard.removeObject(forKey: "vn_apple_userId")
            self?.bridge?.dispatchGlobalEvent("auth:appleCredentialRevoked", payload: [:])
        }
    }

    deinit {
        if let observer = credentialObserver {
            NotificationCenter.default.removeObserver(observer)
        }
    }
}
#endif

File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Modules/SocialAuthModule.kt
package com.vuenative.core

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import androidx.credentials.*
import androidx.credentials.exceptions.GetCredentialCancellationException
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential

/**
 * Native module for social authentication (Google Sign In on Android).
 *
 * Methods:
 *   - signInWithGoogle(clientId: String) -- present Google Sign In via Credential Manager
 *   - signInWithApple() -- not natively supported; returns error with web flow suggestion
 *   - signOut(provider: String) -- clear cached credentials
 *   - getCurrentUser(provider: String) -- check for existing session
 */
class SocialAuthModule : NativeModule {
    override val moduleName = "SocialAuth"
    private var context: Context? = null
    private var bridge: NativeBridge? = null
    private var prefs: SharedPreferences? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun initialize(context: Context, bridge: NativeBridge) {
        this.context = context
        this.bridge = bridge
        this.prefs = context.getSharedPreferences("vn_social_auth", Context.MODE_PRIVATE)
    }

    override fun invoke(method: String, args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        when (method) {
            "signInWithApple" -> {
                // Apple Sign In on Android requires web-based flow
                callback(null, "signInWithApple: not available on Android. Use a web-based Apple OAuth flow instead.")
            }
            "signInWithGoogle" -> {
                val clientId = args.getOrNull(0)?.toString() ?: run {
                    callback(null, "signInWithGoogle: expected clientId string")
                    return
                }
                handleGoogleSignIn(clientId, callback)
            }
            "signOut" -> {
                val provider = args.getOrNull(0)?.toString() ?: "google"
                handleSignOut(provider, callback)
            }
            "getCurrentUser" -> {
                val provider = args.getOrNull(0)?.toString() ?: "google"
                handleGetCurrentUser(provider, callback)
            }
            else -> callback(null, "SocialAuthModule: unknown method '$method'")
        }
    }

    // ── Google Sign In via Credential Manager ───────────────────────────────

    private fun handleGoogleSignIn(clientId: String, callback: (Any?, String?) -> Unit) {
        val ctx = context ?: run { callback(null, "SocialAuth: no context"); return }
        val activity = ctx as? Activity

        if (activity == null) {
            callback(null, "signInWithGoogle: no activity available (context is not an Activity)")
            return
        }

        val credentialManager = CredentialManager.create(ctx)

        val googleIdOption = GetGoogleIdOption.Builder()
            .setFilterByAuthorizedAccounts(false)
            .setServerClientId(clientId)
            .build()

        val request = GetCredentialRequest.Builder()
            .addCredentialOption(googleIdOption)
            .build()

        mainHandler.post {
            kotlinx.coroutines.GlobalScope.launch(kotlinx.coroutines.Dispatchers.Main) {
                try {
                    val result = credentialManager.getCredential(activity, request)
                    val credential = result.credential

                    when (credential) {
                        is CustomCredential -> {
                            if (credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
                                val googleCredential = GoogleIdTokenCredential.createFrom(credential.data)
                                val userInfo = mapOf(
                                    "userId" to googleCredential.id,
                                    "email" to googleCredential.id,
                                    "fullName" to (googleCredential.displayName ?: ""),
                                    "identityToken" to googleCredential.idToken,
                                )

                                prefs?.edit()
                                    ?.putString("google_userId", googleCredential.id)
                                    ?.putString("google_email", googleCredential.id)
                                    ?.putString("google_name", googleCredential.displayName ?: "")
                                    ?.apply()

                                callback(userInfo, null)
                            } else {
                                callback(null, "signInWithGoogle: unexpected credential type")
                            }
                        }
                        else -> callback(null, "signInWithGoogle: unexpected credential type")
                    }
                } catch (e: GetCredentialCancellationException) {
                    callback(null, "signInWithGoogle: user cancelled")
                } catch (e: Exception) {
                    callback(null, "signInWithGoogle: ${e.message}")
                }
            }
        }
    }

    // ── Sign Out ────────────────────────────────────────────────────────────

    private fun handleSignOut(provider: String, callback: (Any?, String?) -> Unit) {
        when (provider) {
            "google" -> {
                prefs?.edit()?.clear()?.apply()
            }
            "apple" -> {
                // No-op on Android
            }
        }
        callback(null, null)
    }

    // ── Get Current User ────────────────────────────────────────────────────

    private fun handleGetCurrentUser(provider: String, callback: (Any?, String?) -> Unit) {
        when (provider) {
            "google" -> {
                val userId = prefs?.getString("google_userId", null)
                if (userId != null) {
                    val result = mapOf(
                        "userId" to userId,
                        "email" to (prefs?.getString("google_email", "") ?: ""),
                        "fullName" to (prefs?.getString("google_name", "") ?: ""),
                    )
                    callback(result, null)
                } else {
                    callback(null, null)
                }
            }
            "apple" -> {
                // Apple Sign In not natively supported on Android
                callback(null, null)
            }
            else -> callback(null, null)
        }
    }
}
